qiskit/qasm3/grammar.py

```diff
@@ -0,0 +1,644 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2021.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+# pylint: disable=invalid-name, abstract-method, super-init-not-called
+
+"""QASM3 AST Nodes"""
+
+
+class Class:
+    """Base abstract class for AST notes"""
+
+    def qasm(self):
+        """Unparses the node"""
+        raise NotImplementedError(self)
```

Comment:
It looks like sometimes `qasm` will return a `str` and sometimes ` list[str]`. Is there a way to know which should be expected?

Reply:
depends if the returning QASM code is a line or not. The program is a tree of lists where the leaves are lines. nested tree means the possibility of indentation. This needs some explanation indeed.

Reply:
Solved in 1ucian0#8